libraries/AP_MSP/AP_MSP_Telem_DJI.cpp

```diff
@@ -0,0 +1,141 @@
+*/
+
+/*
+    The DJI Air Unit polls the FC for the following MSP messages at around 4Hz.
+    Note: messages are polled in ascending hex id order.
+    Hex|Dec|Name
+    ---------------------------
+    03	03	MSP_FC_VERSION
+    0a	10	MSP_NAME
+    54	84	MSP_OSD_CONFIG
+    5c	92	MSP_FILTER_CONFIG
+    5e	94	MSP_PID_ADVANCED
+    65	101	MSP_STATUS
+    69	105	MSP_RC
+    6a	106	MSP_RAW_GPS
+    6b	107	MSP_COMP_GPS
+    6c	108	MSP_ATTITUDE
+    6d	109	MSP_ALTITUDE
+    6e	110	MSP_ANALOG
+    6f	111	MSP_RC_TUNING
+    70	112	MSP_PID
+    82	130	MSP_BATTERY_STATE
+    86	134	MSP_ESC_SENSOR_DATA
+    96	150	MSP_STATUS_EX
+    f7	247	MSP_RTC
+*/
+
+#include <AP_Common/AP_FWVersion.h>
+#include <AP_Vehicle/AP_Vehicle_Type.h>
+#include "AP_MSP.h"
+#include "AP_MSP_Telem_DJI.h"
+
+//#define ENABLE_RESCUE_MODE_DISPLAY
+#if HAL_MSP_ENABLED
+extern const AP_HAL::HAL& hal;
+
+AP_MSP_Telem_DJI::AP_MSP_Telem_DJI(AP_MSP& msp, uint8_t protocol_instance, bool scheduler_enabled) : AP_MSP_Telem_Backend(msp, protocol_instance, scheduler_enabled)
+{
+}
+
+AP_MSP_Telem_DJI::~AP_MSP_Telem_DJI(void)
+{
+}
+
+uint32_t AP_MSP_Telem_DJI::get_osd_flight_mode_bitmask(void)
+{
+    uint32_t mode_mask = 0;
+    // default is hide the DJI flightmode widget
+    BIT_SET(osd_hidden_items_bitmask, OSD_FLYMODE);
+    const AP_Notify& _notify = AP::notify();
+    bool force_visible = false;
+
+    // set arming status
+    if (_notify.flags.armed) {
+        BIT_SET(mode_mask, DJI_MODE_ARM);
+    }
+
+    // check failsafe
+    if (_notify.flags.failsafe_battery || _notify.flags.failsafe_gcs || _notify.flags.failsafe_radio || _notify.flags.ekf_bad ) {
+        BIT_SET(mode_mask, DJI_MODE_FS);
+        force_visible = true;
+    }
+
+#ifdef ENABLE_RESCUE_MODE_DISPLAY
+#if APM_BUILD_TYPE(APM_BUILD_ArduPlane)
+    switch (AP::notify().flags.flight_mode) {
+        case 11:    // RTL
+        case 21:    // QRTL
+            BIT_SET(mode_mask, DJI_MODE_RESC);
+            force_visible = true;
+            break;
+        default:
+            break;
+    }
+#endif
+#if APM_BUILD_TYPE(APM_BUILD_ArduCopter)
+    switch (AP::notify().flags.flight_mode) {
+        case 6:     // RTL
+        case 21:    // SMART RTL
+            BIT_SET(mode_mask, DJI_MODE_RESC);
+            force_visible = true;
+            break;
+        default:
+            break;
+    }
+#endif
+#if APM_BUILD_TYPE(APM_BUILD_Rover)
+    switch (AP::notify().flags.flight_mode) {
+        case 11:     // RTL
+        case 12:     // SMART RTL
+            BIT_SET(mode_mask, DJI_MODE_RESC);
+            force_visible = true;
+            break;
+        default:
+            break;
+    }
+#endif
+#endif //ENABLE_RESCUE_MODE_DISPLAY
+    // do we need to show the DJI flight mode?
+    if (force_visible) {
+        BIT_CLEAR(osd_hidden_items_bitmask, OSD_FLYMODE);
+    }
+    return mode_mask;
+}
+
+bool AP_MSP_Telem_DJI::msp_process_out_api_version(sbuf_t *dst)
```

Comment:
yes but is a failsafe because DJI OSD might require a specific api version to operate (it did in the past), this lets us have a dedicated api+fc version for each backend type